libkbfs/tlf_journal_test.go

```diff
@@ -466,6 +476,85 @@ func testTLFJournalSecondBlockOpWhileBusy(t *testing.T, ver MetadataVer) {
 	putBlock(ctx, t, config, tlfJournal, []byte{1, 2, 3, 4, 5})
 }
 
+func testTLFJournalBlockOpDiskLimit(t *testing.T, ver MetadataVer) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t, ver, TLFJournalBackgroundWorkPaused)
+	defer teardownTLFJournalTest(
+		tempdir, config, ctx, cancel, tlfJournal, delegate)
+
+	tlfJournal.diskLimiter.ForceAcquire(math.MaxInt64)
+	tlfJournal.diskLimiter.Release(6)
+
+	putBlock(ctx, t, config, tlfJournal, []byte{1, 2, 3, 4})
+
+	errCh := make(chan error, 1)
+	go func() {
+		data2 := []byte{5, 6, 7}
+		id, bCtx, serverHalf := config.makeBlock(data2)
+		errCh <- tlfJournal.putBlockData(
+			ctx, id, bCtx, data2, serverHalf)
+	}()
+
+	numFlushed, rev, err := tlfJournal.flushBlockEntries(ctx, 1)
+	require.NoError(t, err)
+	require.Equal(t, 1, numFlushed)
+	require.Equal(t, rev, MetadataRevisionUninitialized)
+
+	// Fake an MD flush.
+	err = tlfJournal.doOnMDFlush(ctx, nil)
+
+	select {
+	case err := <-errCh:
+		require.NoError(t, err)
+	case <-ctx.Done():
+		t.Fatal(ctx.Err())
+	}
+}
+
+func testTLFJournalBlockOpDiskLimitTimeout(t *testing.T, ver MetadataVer) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t, ver, TLFJournalBackgroundWorkPaused)
+	defer teardownTLFJournalTest(
+		tempdir, config, ctx, cancel, tlfJournal, delegate)
+
+	tlfJournal.diskLimiter.ForceAcquire(math.MaxInt64)
+	config._diskLimitTimeout = 3 * time.Microsecond
+
+	data := []byte{1, 2, 3, 4}
+	id, bCtx, serverHalf := config.makeBlock(data)
+	err := tlfJournal.putBlockData(ctx, id, bCtx, data, serverHalf)
+	timeoutErr, ok := errors.Cause(err).(ErrDiskLimitTimeout)
+	require.True(t, ok)
+	require.Error(t, timeoutErr.err)
+	timeoutErr.err = nil
+	require.Equal(t, ErrDiskLimitTimeout{
+		3 * time.Microsecond, int64(len(data)), 0, nil,
+	}, timeoutErr)
+}
+
+func testTLFJournalBlockOpDiskLimitPutFailure(t *testing.T, ver MetadataVer) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t, ver, TLFJournalBackgroundWorkPaused)
+	defer teardownTLFJournalTest(
+		tempdir, config, ctx, cancel, tlfJournal, delegate)
+
+	tlfJournal.diskLimiter.ForceAcquire(math.MaxInt64)
+	tlfJournal.diskLimiter.Release(6)
+
+	data := []byte{1, 2, 3, 4}
+	id, bCtx, serverHalf := config.makeBlock(data)
+	bCtxBad := kbfsblock.MakeContext(
+		keybase1.MakeTestUID(1), keybase1.MakeTestUID(2),
+		kbfsblock.ZeroRefNonce)
+	err := tlfJournal.putBlockData(ctx, id, bCtxBad, data, serverHalf)
+	require.True(t, strings.HasPrefix(err.Error(), "Can't Put() a block"))
```

Comment:
Maybe let the Put fail while verifying the ID, so you can check an actual error type instead of this prefix?